File: Stack and Queue/Parenthesis Checker/checker.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution {
 public:
  bool ispar(string x) {
    stack<char> paren;
    bool flag = true;
    int n = x.length();
    paren.push(x[0]);
    for (int i = 1; i < n; ++i) {
      if (x[i] == ')' || x[i] == '}' || x[i] == ']') {
        if (paren.empty()) {
          flag = false;
          break;
        } else if ((x[i] == ')' && paren.top() != '(') ||
                   (x[i] == '}' && paren.top() != '{') ||
                   (x[i] == ']' && paren.top() != '[')) {
          flag = false;
          break;
        } else {
          paren.pop();
        }
      } else {
        paren.push(x[i]);
      }
    }

    return flag && paren.empty();
  }
};

int main() {
  int t;
  string a;
  cin >> t;
  while (t--) {
    cin >> a;
    Solution obj;
    if (obj.ispar(a))
      cout << "balanced" << endl;
    else
      cout << "not balanced" << endl;
  }
}

File: Stack and Queue/Parenthesis Checker/Checker.java
import java.util.Scanner;
import java.util.Stack;

class Checker {
  public static void main(String args[]) {
    Scanner sc = new Scanner(System.in);
    int t = sc.nextInt();
    while (t-- > 0) {
      String st = sc.next();
      if (new Solution().ispar(st) == true) System.out.println("balanced");
      else System.out.println("not balanced");
    }
    sc.close();
  }
}

class Solution {
  boolean ispar(String x) {
    Stack<Character> paren = new Stack<>();
    boolean flag = true;
    int n = x.length();
    paren.push(x.charAt(0));
    for (int i = 1; i < n; ++i) {
      if (x.charAt(i) == ')' || x.charAt(i) == '}' || x.charAt(i) == ']') {
        if (paren.isEmpty()) {
          flag = false;
          break;
        } else if ((x.charAt(i) == ')' && paren.peek() != '(')
            || (x.charAt(i) == '}' && paren.peek() != '{')
            || (x.charAt(i) == ']' && paren.peek() != '[')) {
          flag = false;
          break;
        } else {
          paren.pop();
        }
      } else {
        paren.push(x.charAt(i));
      }
    }
    return flag && paren.isEmpty();
  }
}
